clippy_lints/src/non_canonical_impls.rs

```diff
@@ -194,6 +194,25 @@ impl LateLintPass<'_> for NonCanonicalImpls {
                 // Fix #11178, allow `Self::cmp(self, ..)` too
                 && self_cmp_call(cx, cmp_expr, impl_item.owner_id.def_id, &mut needs_fully_qualified)
             {
+            }
+            // Fix #12683, allow [`needless_return`] here
+            else if block.expr.is_none()
+                && let Some(stmt) = block.stmts.first()
+                && let rustc_hir::StmtKind::Semi(Expr {
+                    kind: ExprKind::Ret(Some(Expr { kind: ret_kind, .. })),
+                    ..
+                }) = stmt.kind
+                && let ExprKind::Call(
```

Comment:
can this code and the previous code be shared somehow? perhaps a factored out "`expr_is_cmp`" or something

Reply:
Done.